docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,112 @@
+open: func(name: string) -> expected<store, error>
+
+// Get the value associated with the specified `key` from the specified
+// `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store, and `error::no-such-key` will be raised if there is no
+// tuple for `key` in `store`.
+get: func(store: store, key: string) -> expected<list<u8>, error>
+
+// Set the `value` associated with the specified `key` in the specified
+// `store`, overwriting any existing value.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.
+set: func(store: store, key: string, value: list<u8>) -> expected<unit, error>
+
+// Delete the tuple with the specified `key` from the specified `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.  No error is raised if a tuple did not previously
+// exist for `key`.
+delete: func(store: store, key: string) -> expected<unit, error>
+
+// Return whether a tuple exists for the specified `key` in the specified
+// `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.
+exists: func(store: store, key: string) -> expected<bool, error>
+
+// Close the specified `store`.
+//
+// This has no effect if `store` is not a valid handle to an open store.
+close: func(store: store)
+```
+
+*Note: the pseudo-resource design was inspired by the interface of similar functions in [WASI preview 2](https://github.com/bytecodealliance/preview2-prototyping/blob/d56b8977a2b700432d1f7f84656d542f1d8854b0/wit/wasi.wit#L772-L794).*
+
+#### Built-in local key-value database
+
+Spin will have a built-in database based on SQLite for testing, development, and some production use cases.  We’ve selected SQLite for this purpose because it’s easily embeddable, lightweight, and reliable.  It increases the Spin binary size by about 3% and the release build time by about 1.5%.  We’ve also considered lighter-weight options such as an in-memory hash map which is synced to a flat file on updates, but making this reliably atomic and durable would require reinventing the features SQLite already provides.
+
+## Future work
+
+In addition to the built-in, SQLite-based implementation described above, we expect to add a number of other implementations backed by e.g. Redis, other relational databases, eventually consistent distributed stores, etc.  Each of these implementations will have its own performance, consistency, and durability characteristics, and some applications may use a combination of them to handle different types of data.  As mentioned above, this will require defining and configuring named stores in a runtime configuration file.
+
+We also expect to add additional interfaces for e.g. atomic, bulk, and asynchronous operations, key expiration, etc., following [`wasi-keyvalue`](https://github.com/WebAssembly/wasi-keyvalue) as much as possible.
```

Comment:
Would we have a new SIP that defines adding the new interfaces?

Reply:
Perhaps, although if we're just adding features from the `wasi-keyvalue` proposal there might not be much to discuss, in which case a GitHub issue followed by a PR might suffice in lieu of a full-blown SIP.